### PR TITLE
Merge firewall routes

### PR DESCRIPTION
**Pull request by jeremyevans**
@furkansahin I'm asking you to review this because I discovered when merging the routes that we are  using different authorization for the web and api routes.  This applies both authorizations when the specs pass when doing so.  The attach/detach api subnet routes failed if I applied the same authorization used on the web routes to them, so this currently has the api and web attach/detach subnet routes use different authorization.  We don't want that, but I don't know what authorization we do want.  Can you review these commits and determine what authorization we actually want:

* https://github.com/ubicloud/ubicloud/commit/d067c9db90debbd2ce834e6ed81e68e8374a8323
* https://github.com/ubicloud/ubicloud/commit/0322af05307926442a33b1d29796c7edac64e4d5
* https://github.com/ubicloud/ubicloud/commit/548b0a3f69ed5c4e576bdfbd06962c0091d4c412
* https://github.com/ubicloud/ubicloud/commit/5f4871437120f55878c25eb15602df88828dd6ae (didn't test if attach-subnet authorization issue also affects detach-subnet)

Additionally, I found a couple cases that did `response.status = 404` before redirecting.  This doesn't have an effect (redirecting overrides to a 30x status), so I removed it.  However, the flash messages say not found, so maybe it is better to treat these as 404 responses and not redirect.  Please let me know which approach is preferable:

* https://github.com/ubicloud/ubicloud/commit/a3dec8e18ee6e6a15c4ab781444cbe0cc498109b

### CODE REVIEW

File: routes/helpers/firewall.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  def firewall_post(firewall_name)
+    Authorization.authorize(current_account.id, "Firewall:create", @project.id)
+    Validation.validate_name(firewall_name)
+    @location ||= LocationNameConverter.to_internal_name(request.params["location"])
+
+    description = if api?
+      Validation.validate_request_body(request.body.read, [], ["description"])["description"] || ""
+    else
+      request.params["description"]
+    end
```
Comment (enescakir) [inline routes/helpers/firewall.rb:31]: I think we can merge them using json_params similar to vms, private subnets
https://github.com/ubicloud/ubicloud/blob/e80bd6e062b6474a707774d7a33cd50e0db4352a/routes/helpers/vm.rb#L27-L31

File: routes/helpers/firewall.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    {
+      items: Serializers::Firewall.serialize(result[:records]),
+      count: result[:count]
+    }
+  end
+
+  def firewall_post(firewall_name)
+    Authorization.authorize(current_account.id, "Firewall:create", @project.id)
+    Validation.validate_name(firewall_name)
+    @location ||= LocationNameConverter.to_internal_name(request.params["location"])
```
Comment (enescakir) [inline routes/helpers/firewall.rb:25]: In vm and private subnet web routes, we set `@location` in the route. Should we keep it consistent.
https://github.com/ubicloud/ubicloud/blob/190cc6e5706317bd3930398c6702c3d931c2cfa6/routes/merged/project/private_subnet.rb#L11
https://github.com/ubicloud/ubicloud/blob/190cc6e5706317bd3930398c6702c3d931c2cfa6/routes/merged/project/vm.rb#L18

File: routes/merged/project/location/firewall.rb [hunk 3]
```diff
@@ -48,18 +48,15 @@ class Clover
       end
 
       request.delete true do
-        if api?
-          Authorization.authorize(user.id, "Firewall:delete", @project.id)
-
-          firewall.destroy
+        Authorization.authorize(current_account.id, "Firewall:delete", @project.id)
+        Authorization.authorize(current_account.id, "Firewall:delete", firewall.id)
```
[the remark this answers is not in the document]
Comment (jeremyevans) [reply routes/merged/project/location/firewall.rb]: Pushed d41e902d30bf2ba481782fcbf70603b941be9271 for this and other cases.

File: routes/merged/project/location/firewall.rb [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+          if api?
+            fail Validation::ValidationFailed.new({private_subnet_id: "Private subnet with the given id \"#{private_subnet_id}\" and the location \"#{@location}\" is not found"})
+          else
+            flash["error"] = "Private subnet not found"
+            r.redirect "#{@project.path}#{firewall.path}"
+          end
+        end
+
+        # XXX: differing authorization between api and web routes!
+        Authorization.authorize(current_account.id, "PrivateSubnet:edit", private_subnet.id) unless api?
```
[the remark this answers is not in the document]
Comment (jeremyevans) [reply routes/merged/project/location/firewall.rb:77]: This requires changes to the specs, which @furkansahin said he would work on.

[the remark this answers is not in the document]
Comment (jeremyevans) [reply routes/merged/project/location/firewall.rb]: I fixed the specs and rebased this on main (833ae470485756189a76dd8d089a95d16ef23286).  I also made the validation the same for both api and web routes (d9f13569249280f31d9bb320b56ac49a32b1f1d4).

File: routes/merged/project/location/firewall.rb [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+          @attachable_subnets = Serializers::PrivateSubnet.serialize(project_subnets.reject { |ps| attached_subnets.find { |as| as.id == ps.id } })
+
+          view "networking/firewall/show"
+        end
+      end
+
+      r.post %w[attach-subnet detach-subnet] do |action|
+        Authorization.authorize(current_account.id, "Firewall:view", firewall.id)
+
+        key = api? ? "private_subnet_id" : "private-subnet-id"
```
Comment (byucesoy) [inline routes/merged/project/location/firewall.rb:59]: I think we can rename the field in our web pages to `private_subnet_id` and eliminate this branch. This is what I did for previous merges into `Common`

Comment (jeremyevans) [reply routes/merged/project/location/firewall.rb]: That makes sense to me.  I added a pull request for that.

**Review by furkansahin — APPROVED** [review]

**Pull request merged**